### ISSUE TITLE
Unexpected behavior with linked NPM modules

### ISSUE THREAD

**Issue by ooflorent** [webpack/webpack#985]
## tl;dr

Webpack should produce the same build whether a module is linked or not.
## In depth

Considering the following dependency trees:

``` js
project@1.0.0
├── babel-runtime@5.1.9
└── sub-project@1.0.0
```

``` js
sub-project@1.0.0
└── babel-runtime@5.1.9
```

**Notice:** NPM does not include `babel-runtime` under `sub-project` when `npm list` is called from `project`.

Webpack builds something link:

``` js
[0] multi main 52 bytes {0} [built]
[1] ./index.js 2.08 kB {0} [built]
[2] ./~/babel-runtime/helpers/interop-require-wildcard.js 148 bytes {0} [built]
[3] ./~/sub-project/index.js 2.46 kB {0} [built]
```

If we run `npm link sub-project` we obtain the following tree:

``` js
project@1.0.0
├── babel-runtime@5.1.9
└── sub-project@1.0.0 -> /Users/fc/Development/sub-project
```

**Notice:** `babel-runtime` still doesn't appear under `sub-project`.

But building the project again will duplicate its dependencies:

``` js
[0] multi main 52 bytes {0} [built]
[1] ./index.js 2.08 kB {0} [built]
[2] ./~/babel-runtime/helpers/interop-require-wildcard.js 148 bytes {0} [built]
[3] /Users/fc/Development/sub-project/index.js 2.46 kB {0} [built]
[4] /Users/fc/Development/sub-project/~/babel-runtime/helpers/interop-require-wildcard.js 148 bytes {0} [built]
```

Using `resolve.root` or changing the resolve scope of loaders do not change anything.
## Expectations

It would link to be able the link my in-development modules into my project using `npm link` without having to tweak the webpack config.

**The real problem isn't the build size but the unexpected behaviors raised but duplicated dependencies.**

**Comment by ooflorent** [comment]
See #943 for related discussions. cc @sokra

**Comment by jhnns** [comment]
Mhmmm ... we've discussed about that already but I can't find the issue anymore. The problem is that linked modules have another resolved filename and I'm not sure if we can just change that. For instance, node's `__filename` also resolves to the real path...

**Comment by ooflorent** [comment]
Problem is React uses some global definitions and static objects (such as `ReactInstanceMap`). When included twice, React is unable to figure out if the component exists. In some complex cases using `context` it totally fails.

**Comment by sokra** [comment]
I see the problem, but I don't know how to solve it. webpack implements the node.js resolving algorithm: https://nodejs.org/api/modules.html#modules_all_together
This describes how modules are handled and which places should be looked up for dependencies (that's not the application node_modules folder).

So the problem is by design and should occur in node.js too.

A workaround is to let webpack prefer the application node_modules with: `resolve.root` and `resolveLoader.root` = `path.join(__dirname, "node_modules")`

**Comment by ooflorent** [comment]
Unfortunately settings `resolve.root` and `resolveLoader.root` does not solve the issue.

```
Hash: 6d255d7e6eb942c864e9
Version: webpack 1.8.4
Time: 3896ms
                    Asset       Size  Chunks             Chunk Names
               index.html  345 bytes          [emitted]
                  main.js    4.71 MB       0  [emitted]  main
chunk    {0} main.js (main) 1.67 MB [rendered]
    [0] multi main 52 bytes {0} [built]
...
  [478] ./~/react/lib/toArray.js 2.08 kB {0} [built]
...
  [522] /Users/fc/Development/ooflorent/xxxxxx/~/react/lib/toArray.js 2.08 kB {0} [built]
webpack: bundle is now VALID.
```

**Comment by nevir** [comment]
If you're loading babel via `babel-loader`, you can add `include: __dirname, exclude: /node_modules/` to its configuration (assuming `__dirname` matches your project dir)

**Comment by johnelliott** [comment]
I am having this issue too. At some point my changes to my module stopped rebuilding the webpack bit of a project I was integrating it with.

**Comment by jasonslyvia** [comment]
Hate to bother but is there any update about this issue?

**Comment by MoOx** [comment]
the workaround explained using `npm link` to get one version of react only (on both webpack and node) helped me https://github.com/webpack/webpack/issues/966#issuecomment-95491120

**Comment by johnelliott** [comment]
@jasonslyvia I got around this too that same day. I had to gut the linking and dump every cache and node module I could find. I still don't know if the root of the issue was this code or my fault.

**Comment by jessepollak** [comment]
I have this same issue — the fix that @MoOx specifies works, but I have packages with a bunch of shared dependencies, so it's not a viable long term solution.

**Comment by chrisbateman** [comment]
Preferring the application node_modules with `resolveLoader.root` worked well enough with npm 2 - but with npm 3 it means that you will randomly pull in the wrong version of a package when there are duplicate versions within the app.

For instance - say my app uses moduleA which depends on lodash@3 and moduleB which depends on lodash@4. The npm3 install ends up looking like this:
- lodash@3
- moduleA
- moduleB
  - lodash@4

If I set `resolveLoader.root` - moduleB ends up pulling in lodash 3 rather than 4 (since lodash@3 just happened to get installed at root).

**Comment by jure** [comment]
So for this particular problem, there are now two useful solutions (but quite tricky ones): 

1. npm linking the linked module's dependencies to the `project`'s dependency, i.e. https://github.com/webpack/webpack/issues/966#issuecomment-95491120

2. sharing a `node_modules` folder between linked module and `project`, i.e.
https://github.com/webpack/webpack/issues/554#issuecomment-203128656

I don't like either of those due to their manual/repetitive/unstable nature. I'd prefer this solution: https://github.com/aurelia/webpack-plugin/issues/44#issuecomment-238028171, a webpack ResolverPlugin that replaces the linked modules node_modules requires with the `project`'s. That's what we're attempting. Any other solutions out there?

**Comment by ganmor** [comment]
One possible solutions is not using npm link at all.
If you are not using window, here is a nice alternative https://github.com/wix/wml

**Comment by lazd** [comment]
Uhg this issue is plaguing me right now. The second I link a package in that has a dependency in common with the parent project, my build contains duplicates.

**Comment by niieani** [comment]
I've written a resolve plugin which takes care of this issue.

What is does is tries to resolve any module calls to the closest one to the provided `rootDir`, always preferring local ones, as long as the version of the closer dependency is within the requested package's SemVer range. Useful especially when using linked modules, but even when not using linked modules it it's pretty useful, as you don't end up with duplicate copies of the same package if another copy is within a nested `node_module`.

It's called [RootMostResolvePlugin](https://github.com/niieani/webpack-dependency-suite/blob/master/plugins/root-most-resolve-plugin.ts) and it's a part of the [webpack-dependency-suite](https://github.com/niieani/webpack-dependency-suite) (sorry, no docs yet).

Configuration is [pretty simple](https://github.com/aurelia/skeleton-navigation/blob/3eb4d3fa65b60731493c79235e547792324cc3e7/skeleton-typescript-webpack-experimental/webpack.config.ts#L367-L370).

Make sure to put the plugin in the `resolve` object of your configuration.

When instantiating, the first parameter should be the root directory of your project (or the one in which context you want to resolve) and the second, optional parameter is whether you want to enforce it for all paths (e.g. including linked modules, use `true`), or only for nested `node_modules` (`false`).

**Comment by jure** [comment]
Hey, just a quick update on our side. We've managed to solve our problems using `npm link`ed dependencies by:

Using Webpack 2 and setting 

```
  resolve: { 
      symlinks: false,
      modules: [
        path.resolve(__dirname, '..', 'node_modules'),
        'node_modules'
      ],
  }
```

So the local/root node_modules is always preferred, and symlinks are not resolved, i.e. same as the node `--preserve-symlinks` options, this was implemented in: https://github.com/webpack/webpack/issues/2937

This is working as expected for us, if that helps anyone stumbling upon this issue.

**Comment by niieani** [comment]
@jure, you're opening your project up to potential problems if you get many major/minor versions of the same package dependent on a different one, because `resolve.modules` does not verify the version of the package. If you want to mitigate that and ensure you are only including modules that match the requesting module's SemVer constraint, you can use the resolve plugin I mentioned before.

**Comment by bebraw** [comment]
Can someone check this against webpack 2? Thanks.

**Comment by niieani** [comment]
@bebraw The problem described in the issue is still present in Webpack 2, and as far as I understand will not be fixed, because the dependency resolution algorithm is based on the one provided by Node. Webpack works properly, but it's an arguably bad design decision on Node's part that Webpack follows. A webpack resolve plugin can be used to mitigate the problem successfully.

**Comment by travi** [comment]
just to provide a bit more insight of what is working, i was able to get this working in webpack 2 with a similar `resolve` config as above. thanks @jure!

```js
resolve: { 
  symlinks: false,
  modules: [path.resolve('node_modules')],
}
```

it seems like it would pretty safe to default to the local `node_modules/` without needing to configure `resolve.modules`, maybe based on `context`?

**Comment by niieani** [comment]
@travi not at all safe. If you have `lodash` version 4 in `node_modules` and a package requires `lodash` version 2, it will break with `modules: [path.resolve('node_modules')]`. You should use `RootMostResolvePlugin` as it does essentially what `modules: [path.resolve('node_modules')]` does but also does a `semver` check, to see if the packages are compatible.

**Comment by travi** [comment]
@niieani i understand your point, but if i had a situation like that, i would be wanting to track it down and eliminate it anyway. having multiple versions like that would result in multiple version included and inflating the bundle, so even when not linking, i would want to enable deduping those situations.

**Comment by Undistraction** [comment]
@niieani Thanks. Your plugin fixes this nicely npm@4.3.0, webpack@2.2.1

**Comment by zinserjan** [comment]
@niieani I tried to use your plugin with a lerna based repo but unfortunately it does not work for me (webpack@2.2.1).

I added several logs into your plugin and in my case it will never reach the following line https://github.com/niieani/webpack-dependency-suite/blob/f36e40edd64ebc3a045b76fe2735f868bd005b3e/plugins/root-most-resolve-plugin.ts#L40.

The reason for this is that the _issuer_ information was removed in https://github.com/webpack/webpack/commit/da29d21ae44056ccbd525282300df7184e922d97#diff-1fd8fe07e83988b2301f9bad408cf5c2 and therefore it will never reach the code where the magic happens (contextInfo was replaced with an empty object).

To get this working again I had to modify the `NormalModuleFactory.js` inside the `node_modules/webpack/lib` folder and replaced `var resolveContextInfo = {};` with `var resolveContextInfo = contextInfo;`.

This brings me to the question if this needs to be fixed by webpack or can we fix the plugin and access the necessary _issuer_ information in another way?

Maybe @sokra can help us to understand why `resolveContextInfo` was changed to an empty object. I guess that this changed cause issuer in context was never used before in the whole codebase. At least I wasn't able to find any usages of the issuer in the resolving process.

Btw this plugin is pretty cool, especially in combination with [duplicate-package-checker-webpack-plugin](https://github.com/darrenscerri/duplicate-package-checker-webpack-plugin).

**Comment by fabb** [comment]
Won‘t `symlinks: false` break HMR, so when I change something in the linked module, the client project does no longer update via HMR automatically?

**Comment by fabb** [comment]
#811 looks related as well.

**Comment by kleinph** [comment]
I stumbled upon this issue in a monorepo where one package is a dependency of another via `"other-package": "file:../other-package"`.

But this is no webpack issue, the root cause seems to be that npm installs the deps of the sub project in the sub projects _node_modules_ folder instead of the parent projects one (as it is normally the case).

There seems to be a few issues open in npm regrading this [1], [2], [3]. The last one points out, that this behavior is per spec (see: [4]).

But I found another solution for this: using yarn instead of npm will not create the unnecessary _node_modules_ folder in the sub project and install all deps in the parent project.

1: https://npm.community/t/npm-does-not-flatten-packages-during-install/3989/3
2: https://npm.community/t/npm-doesnt-hoist-linked-module-dependencies-to-project-level-when-new-dependencies-are-added-to-linked-module/5717/3
3: https://npm.community/t/npm-doesnt-install-local-package-dependencies/805
4: https://github.com/npm/cli/blob/latest/doc/spec/file-specifiers.md#installation

**Comment by patrick-webs** [comment]
I encountered this issue and I was able to fix it with the `RootMostResolvePlugin` referenced above.

My project was on webpack 1, so I just updated to webpack 2.x and was able to use the plugin. But I'm noticing on the repository that it seems like the plugin isn't available for webpack 4.x and isn't actively developed anymore, is anyone aware of another alternative that's compatible with the newest webpack? https://github.com/niieani/webpack-dependency-suite/issues/31

**Comment by trusktr** [comment]
Although it would stray from the standard Node.js resolution algorithm, I strongly believe Webpack should be smarter here (at least behind an option) and prefer packages found in the app's `node_modules` (the root-most `node_modules`) if those packages meet the semver requirements of the other packages depending on them.

@niieani's RootMostResolvePlugin (above) does exactly this.

@sokra Do you think it would be worth making this behavior the default in Webpack?

**Comment by trusktr** [comment]
@niieani I tried `RootMostResolvePlugin` with Webpack 4.16.4, and I get

```
Error: Plugin could not be registered at 'resolved'. Hook was not found.
```

Does it need an update?

**Comment by goatfryed** [comment]
Hey guys,
I ran into this issue while trying to develop a library feature with an optional dependency. I noticed that my demo project - although not requiring the dependecy - still got it bundled. So i removed the optional dependency from my libs module folder, which broke the second demo project that should have had this dependency.
It took my quite sometime and testing until i finally found this issue and #811.

Currently, this is mentioned in https://webpack.js.org/configuration/resolve/#resolve
What do you guys think about adding a passage in https://webpack.js.org/guides/author-libraries/ as well?

**Comment by vankop** [comment]
Should work with `webpack@5`. Feel free to report new issue with reproducible repo.

**Comment by fabis94** [comment]
@vankop is there a way to get this functionality in webpack@4 as well?

**Comment by vankop** [comment]
no, only critical fixes for webpack@4 is possible..

**Comment by fabis94** [comment]
@vankop are there any docs about how this is supposed to work in webpack@5 or what are the limitations? cause I'm still running into the same issue - three.js is loaded both from the package's own `node_modules` dir and also from a symlinked local packages `viewer` `node_modules` dir (so `./node_modules/three` and `../viewer/node_modules/three`)

**Comment by vankop** [comment]
@fabis94 from your description is unclear what is wrong.. 
> three.js is loaded both from the package's own node_modules dir and also from a symlinked local packages

maybe some dependency ( or different app modules ) is trying to resolve `three.js` and resolve wrong one..

Probably in your case problem is not related to symlinks.. ( it will work same way with non-symlink version ) you can reduce this to reproducible repo I will take a look.

**Comment by vankop** [comment]
I read the description once more.. I thought this related to symlinks..

"Issue" still exists

**Comment by fabis94** [comment]
Just an idea, but why not just compare if the same version of the same package is already bundled in? Or make that an opt-in feature?

The underlying issue is that after linking in modules, npm doesn't deduplicate the node_modules tree so you can end up having `/node_modules/three` and `../symlinked-package/node_modules/three` both available. And then `three` is imported from `symlinked-package` it will use its own local `three` version, but when `three` is imported by the host application built by webpack it'll use `/node_modules/three`, resulting in 2 three.js versions bundled in at the same time.

But this seemingly npm-specific issue is definitely not going away, considering that symlinking packages into node_modules is a very common workflow. So I think it makes sense for the improvement to be introduced on the webpack side, instead.

**Comment by vankop** [comment]
"fix" is possible, but this breaks Node.js resolution algorithm.

**Comment by fabis94** [comment]
I don't really see why people would want the current behaviour (same package with same version being bundled twice), but it could be an opt-in feature, if you think it's gonna cause issues for some.

Anyone trying to deal with these issues and spending hours on it not getting anywhere isn't going to care that the Node.js resolution algorithm is broken, getting dependencies to be properly bundled without duplicates is more important. And of course - by opting in you'd understand what you're doing and the consequences.
